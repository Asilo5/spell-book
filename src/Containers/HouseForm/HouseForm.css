* {
    padding: 0;
    margin: 0;
    box-sizing: border-box;
}

.form-section {
  background-image: url('https://thumbs.gfycat.com/HardHoarseBarasinga-size_restricted.gif');
  background-size: cover;
  background-repeat: no-repeat;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
}

.house-form {
    display: flex;
    justify-content: center;
    flex-direction: column;
    width: 30%;
    margin: 10px auto;
}

.wizard-name {
    font-size: 35px;
    border: none;
    letter-spacing: 2px;
    background: transparent;
    position: absolute;
    top: 1em;
    left: 0.7em;
    text-align: center;
    font-family: 'Henny Penny', cursive;
}

.wizard-name::placeholder {
  font-family: 'Henny Penny', cursive;
}

.banner {
   animation: float 5s linear infinite;
}

.sorting-button,
.sorting-button-invalid {
    width: 18em;
    display: block;
    margin: auto;
    cursor: pointer;
}

.sorting-button-invalid {
  cursor: inherit;
} 

.sorting-container {
  transition: transform .8s ease-in-out;
}

.sorting-container:hover {
    transform: rotate(360deg);
}

.name-banner {
  height: 10em;
  margin-left: -5em;
}

.submit-text, 
.submit-hat {
  position: relative;
  text-align: center;
  bottom: 131px;
  left: 13px;
  color: #f5be77;
  font-size: 37px;
  text-shadow: 2px 2px black;
  font-family: 'ZCOOL XiaoWei', serif;
}

.submit-hat {
  font-size: 37px;
  color: #f11818;
}

a {
  text-decoration: none;
}

@keyframes float {
    20% {
      transform: translate(3px, 1px);
    }
    50% {
      transform: translate(20px, 5px); 
    }
    70% {
      transform: translate(30px, 8px);
    }
    100% {
      transform: translate(3px, 1px);
    }
}