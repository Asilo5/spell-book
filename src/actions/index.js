export const givenHouse = (house) => ({
    type: 'GIVEN_HOUSE',
    house
});

export const isLoading = (bool) => ({
    type: 'IS_LOADING',
    bool
});

export const wizardsName = (name) => ({
  type: 'NAME',
  name
});

export const hasError = (error) => ({
  type: 'HAS_ERROR',
  error
})

export const hasSpells = (spells) => ({
  type: 'HAS_SPELLS',
  spells
})