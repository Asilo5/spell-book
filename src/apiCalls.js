const apiKey = `$2a$10$s7FC7dj4ik0afkB/Dnr4Ju6NXS0Os8RtRntk97bi8gx.KqvmqgNgK`;

export const getHouses = async () => {
    const response = await fetch(`https://www.potterapi.com/v1/sortingHat`);
    const data = await response.json();

    if(!response.ok) {
      throw new Error('No Houses Found!');
    } else {
      return data;
    }
}