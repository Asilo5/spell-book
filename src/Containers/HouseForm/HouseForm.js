import React, { Component } from 'react';
import './HouseForm.css';
import { getHouses, getSpells } from '../../apiCalls';
import { connect } from 'react-redux';
import { givenHouse, isLoading, wizardsName, hasSpells, searchedSpell } from '../../actions';
import { Link } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import LoadingPage from '../LoadingPage/LoadingPage';

export class HouseForm extends Component {
    constructor() {
      super();
      this.state = {
        name : '',
        hasError: ''
      }
    }

    handleChange = (e) => {
      this.setState({ [e.target.name] : e.target.value })
    }

    foundHouse = async () => {
        const { givenHouse, wizardsName, isLoading } = this.props;
        const { name } = this.state;
      try {
        const resp = await getHouses();
        // { isLoading && <LoadingPage /> }
        givenHouse(resp);
        wizardsName(name);
        isLoading(false);
        this.foundSpells();
    } catch (error) {
        this.setState({ hasError : error });
    }
  }

   foundSpells = async () => {
     const { hasSpells, searchedSpell } = this.props;
     try {
       const resp = await getSpells();
       let updateResp = resp.map((spell) => {
         spell.isFavourite = false;
         return spell;
       });
       hasSpells(updateResp);
       searchedSpell(updateResp);
     } catch (error) {
      this.setState({ hasError : error });
     }
   }

render() {
  const { name } = this.state;
  const isEnabled = name.length > 0;
      return (
          <section className='form-section'>
              <form className='house-form'>
                  <div className='banner'>
                    <input className='wizard-name'
                         type='text'
                         placeholder='Name Here' 
                         name='name' 
                         value={this.state.name} 
                         onChange={this.handleChange} />
                    <img className='name-banner' src='https://www.pinclipart.com/picdir/big/10-101144_vintage-banner-vector-png-theveliger-clipart-vintage-banner.png' alt='banner name' />
                  </div>
                  <Link to='/spells' onClick={this.foundHouse} disabled={!isEnabled} > 
                    <div className='sorting-container'>
                      <img className='sorting-button' src='https://cdn.shopify.com/s/files/1/0221/1146/products/Sorting_Hat_pin_badge_scaled_grande.png?v=1551715337' alt='sorting hat' />
                      <h3 className='submit-text'>Enter</h3>
                    </div>
                  </Link>
              </form>
          </section>
      )
    }
}

export const mapDispatchToProps = (dispatch) => (
   bindActionCreators({
     givenHouse: house => dispatch( givenHouse(house) ),
     isLoading: bool => dispatch( isLoading(bool) ),
     wizardsName: name => dispatch( wizardsName(name) ),
     hasSpells: spells => dispatch( hasSpells(spells) ),
     searchedSpell: spell => dispatch( searchedSpell(spell) )
   }, dispatch)
)


export default connect(null, mapDispatchToProps)(HouseForm);