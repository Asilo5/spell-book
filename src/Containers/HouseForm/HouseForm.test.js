import React from 'react';
import { shallow } from 'enzyme';
import { HouseForm, mapDispatchToProps } from './HouseForm';
import { givenHouse, isLoading, wizardsName, hasSpells, searchedSpell } from '../../actions';
import { getHouses, getSpells } from '../../apiCalls';

jest.mock('../../apiCalls');

describe('HouseForm', () => {
    describe('Form', () => {
       
        let wrapper;
        let mockName = 'Bob Potter';
        let mockHouse = 'Hufflepuff';
        let mockLoading = true;
        let mockSpells = [
            { _id: '5b74ebd5fb6fc0739646754c',
               spell: 'Aberto',
               type: 'Charm',
               effect: 'opens objects' }];

        beforeEach(() => {
          
            wrapper = shallow( <HouseForm 
               givenHouse={mockHouse}
               isLoading={mockLoading}
               wizardsName={mockName}
               hasSpells={mockSpells}
               searchedSpell={mockSpells} />);

            getHouses.mockImplementation(() => {
                return Promise.resolve(mockHouse);
            });

            getSpells.mockImplementation(() => {
                return Promise.resolve(mockSpells);
            })
        })

        it('should match snapshot with all information passing through', () => {
            expect(wrapper).toMatchSnapshot();
        })

        it('should update local state of name when handleChange is invoked', () => {
            const mockEvent = {
                target: {
                  name: 'name',
                  value: 'Bob Potter'
                }
            }

            const expected = 'Bob Potter';

            wrapper.instance().handleChange(mockEvent);
            expect(wrapper.state('name')).toEqual(expected);
        })

        it.skip('should invoke findHouse when link is clicked', () => {
            wrapper.instance().foundHouse = jest.fn();

            wrapper.find('Link').simulate('click');
            expect(wrapper.instance().foundHouse).toHaveBeenCalled();
        })

        it('should call getHouse fetch when foundHouse is called', () => {
            wrapper.instance().foundHouse();
            expect(getHouses).toHaveBeenCalled();
        })

        it('should call getSpells when foundSpells is called', () => {
            wrapper.instance().foundSpells();
            expect(getSpells).toHaveBeenCalledWith();
        })
    })

    describe('mapDispatchToProps', () => {
        
            let mockName = 'Bob Potter';
            let mockHouse = 'Hufflepuff';
            let mockLoading = true;
            let mockSpells = [
                { _id: '5b74ebd5fb6fc0739646754c',
                   spell: 'Aberto',
                   type: 'Charm',
                   effect: 'opens objects' 
                }];
            const mockDispatch = jest.fn()
       it('should dispatch givenHouse when foundHouse is called', () => {
           const actionToDispatch = givenHouse(mockHouse);
           const mappedProps = mapDispatchToProps(mockDispatch);

           mappedProps.givenHouse(mockHouse);
           expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
       })

       it('should dispatch wizardsName when foundHouse is called', () => {
          const actionToDispatch = wizardsName(mockName);
          const mappedProps = mapDispatchToProps(mockDispatch);

          mappedProps.wizardsName(mockName);
          expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
       })

       it('should dispatch hasSpells when foundSpells is called', () => {
           
       })
    })
})