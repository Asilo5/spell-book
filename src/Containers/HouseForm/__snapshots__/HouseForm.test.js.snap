// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HouseForm Form should match snapshot with all information passing through 1`] = `
<section
  className="form-section"
>
  <form
    className="house-form"
  >
    <div
      className="banner"
    >
      <input
        className="wizard-name"
        name="name"
        onChange={[Function]}
        placeholder="Name Here"
        type="text"
        value=""
      />
      <img
        alt="banner name"
        className="name-banner"
        src="https://www.pinclipart.com/picdir/big/10-101144_vintage-banner-vector-png-theveliger-clipart-vintage-banner.png"
      />
    </div>
    <Link
      disabled={true}
      onClick={[Function]}
      to="/spells"
    >
      <div
        className="sorting-container"
      >
        <img
          alt="sorting hat"
          className="sorting-button"
          src="https://cdn.shopify.com/s/files/1/0221/1146/products/Sorting_Hat_pin_badge_scaled_grande.png?v=1551715337"
        />
        <h3
          className="submit-text"
        >
          Enter
        </h3>
      </div>
    </Link>
  </form>
</section>
`;
